### PR TITLE
Adding support for silent option to download

### PR DESCRIPTION
**Pull request by qyra**
Adding support for the silent=Bool keyword when calling download.
The default behaviour is to display progress indicators in wget and curl, to match the previous behaviour of download(). Passing silent=true will hide output from wget and curl.
This is primarly aimed at making download usable in async environments, without causing jumbled output.
The relevant documentation is also updated to note the new parameter, and that windows URLs must have the http:// suffix (wget / curl adds it automatically)

A previous pull request was made against the release-0.4 branch accidentally - see https://github.com/invenia/julia/pull/2 All comments from there should be addressed in this

### CODE REVIEW

File: base/docs/helpdb/Base.jl [hunk 1]
```diff
@@ -130,13 +130,15 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
 takebuf_array
 
 """
-    download(url,[localfile])
+    download(url,[localfile]; silent=false)
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: needs rst signature update. and tests

**Comment by tkelman** [comment]
x-ref a previous attempt at https://github.com/JuliaLang/julia/pull/12149, I think this is potentially a bit cleaner though

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
 takebuf_array
 
 """
-    download(url,[localfile])
+    download(url [localfile] [silent=false])
```
Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: Why did you remove the comma? OTC, you need to add one before `silent`. Also, since there is `=false`, the brackets are not needed.

Same below.

File: base/docs/helpdb/Base.jl [hunk 3]
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
...
 
 Download a file from the given url, optionally renaming it to the given local file name.
+Passing silent=true will suppress progress indicators.
```
Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: Needs backquotes around `silent=true`.

File: base/docs/helpdb/Base.jl [hunk 4]
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
...
 Note that this function relies on the availability of external tools such as `curl`, `wget`
 or `fetch` to download the file and is provided for convenience. For production use or
 situations in which more options are needed, please use a package that provides the desired
-functionality instead.
+functionality instead. Platform specific details: In windows, download requires a full URL,
+complete with the http:// prefix.
+"""
+download
+
+"""
```
Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: This is redundant with the docstring above, since `[localfile]` implies that argument can be omitted.

File: base/docs/helpdb/Base.jl [hunk 5]
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
...
-    download(url,[localfile])
+    download(url [localfile] [silent=false])
 
 Download a file from the given url, optionally renaming it to the given local file name.
+Passing silent=true will suppress progress indicators.
 Note that this function relies on the availability of external tools such as `curl`, `wget`
 or `fetch` to download the file and is provided for convenience. For production use or
 situations in which more options are needed, please use a package that provides the desired
-functionality instead.
+functionality instead. Platform specific details: In windows, download requires a full URL,
```
Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: Please move `Platform specific details:` to a separate paragraph, as a title like this: `# Platform-Specific Details`. I would also rephrase it as something like "On Windows, a full URL must be provided, including the protocol prefix (e.g. "http://" for HTTP)."

File: base/docs/helpdb/Base.jl (hunk 2 above)
Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: Ah, and since this is a keyword argument, you need to put `;` before `silent` (instead of a comma, contrary to what I said above).

Comment (akrasian) [inline base/docs/helpdb/Base.jl]: I was having difficulty getting doc generation to work and trying to figure out why it was not matching the doc signatures. Editing the .rst files worked,I should not have pushed those changes. My current push corrects that.

Comment (nalimilan) [inline base/docs/helpdb/Base.jl]: Looks like you didn't push the changes.

**Comment by akrasian** [comment]
Currently working on tests for this. There are a couple strange things to deal with because redirect_stdout() and redirect_stderr() don't actually let you redirect stdout and stderr from system calls for some reason.

**Comment by akrasian** [comment]
Added a test file for interactiveutil which tests download silent/verbose, with or without a specified file, and addressed all comments made so far.

File: test/interactiveutil.jl [hunk 6]
```diff
@@ -0,0 +1,110 @@
+### Testing download(url,[localfile],[silent=false])
+
+"Runs an expression, and returns stdout, stderr, and the result."
+function capture_output(expression)
+	originalSTDOUT = STDOUT
```
Comment (tkelman) [inline test/interactiveutil.jl]: 4-space indent, no tabs

needs a rebase

File: test/choosetests.jl [hunk 7]
```diff
@@ -33,7 +33,7 @@ function choosetests(choices = [])
         "markdown", "base64", "serialize", "functors", "misc", "threads",
         "enums", "cmdlineargs", "i18n", "workspace", "libdl", "int",
         "checked", "intset", "floatfuncs", "compile", "parallel", "inline",
-        "boundscheck", "error"
+        "boundscheck", "error", "interactiveutil"
```
Comment (tkelman) [inline test/choosetests.jl]: will need to be added to the `net_required_for` array below

File: test/interactiveutil.jl [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	redirect_stderr(originalSTDERR)
+	redirect_stdout(originalSTDOUT)
+
+	errstr = join(map(x->Char(x), errors))
+	outstr = join(map(x->Char(x), output))	
+	# Remove the manually added \n from stdout and stderr
+	return (chop(outstr), chop(errstr), result)
+end
+
+port = 32744 #TODO - find generic way to get an unused port.
```
Comment (omus) [inline test/interactiveutil.jl]: What you are looking for is [`listenany`](http://julia.readthedocs.org/en/latest/stdlib/io-network/#Base.listenany)

```julia
port, server = listenany(2000)
```

File: test/interactiveutil.jl [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+Content-Type: text/html
+Connection: Closed
+
+$html"""
+
+function start_server()
+	server_running = true
+	@async begin
+		server = listen(port)
+		while server_running
```
Comment (omus) [inline test/interactiveutil.jl]: You need to declare `server_running` as global on [line 43](https://github.com/JuliaLang/julia/pull/15930/files#diff-125a8ef99a2617e382db614c9a77673fR43) and [line 56](https://github.com/JuliaLang/julia/pull/15930/files#diff-125a8ef99a2617e382db614c9a77673fR56). You can also delete [line 33](https://github.com/JuliaLang/julia/pull/15930/files#diff-125a8ef99a2617e382db614c9a77673fR33)

File: test/interactiveutil.jl [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	end
+end
+
+function stop_server()
+	server_running = false
+end
+
+downloader = :unknown
+for checkcmd in (:curl, :wget, :fetch)
+	if success(pipeline(`which $checkcmd`, DevNull))
```
Comment (omus) [inline test/interactiveutil.jl]: `success` won't display STDOUT from the command. You can safely remove `pipeline` here.

File: test/interactiveutil.jl [hunk 11]
```diff
@@ -0,0 +1,110 @@
+### Testing download(url,[localfile],[silent=false])
+
+"Runs an expression, and returns stdout, stderr, and the result."
+function capture_output(expression)
```
Comment (omus) [inline test/interactiveutil.jl]: Could avoid using `eval` if you would change the signature take in a function and its arguments:
```julia
function capture_output(func::Function, args...; kwargs...)
    ...
    result = func(args...; kwargs...)
    ...
end

```

File: test/interactiveutil.jl [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	originalSTDOUT = STDOUT
+	originalSTDERR = STDERR
+	(errRead, errWrite) = redirect_stderr()
+	(outRead, outWrite) = redirect_stdout()
+
+	result = eval(expression)
```
Comment (omus) [inline test/interactiveutil.jl]: Should be in a `try` block where the finally block restores the original STDOUT/STDERR

File: test/interactiveutil.jl [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	output = readavailable(outRead)
+	close(outRead)
+	close(errRead)
+	redirect_stderr(originalSTDERR)
+	redirect_stdout(originalSTDOUT)
+
+	errstr = join(map(x->Char(x), errors))
+	outstr = join(map(x->Char(x), output))	
+	# Remove the manually added \n from stdout and stderr
+	return (chop(outstr), chop(errstr), result)
```
Comment (omus) [inline test/interactiveutil.jl]: Should be using [`chomp`](http://docs.julialang.org/en/latest/stdlib/strings/#Base.chomp) rather than [`chop`](http://docs.julialang.org/en/latest/stdlib/strings/#Base.chomp) here

File: test/interactiveutil.jl [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	close(outWrite)
+	errors = readavailable(errRead)
+	output = readavailable(outRead)
+	close(outRead)
+	close(errRead)
+	redirect_stderr(originalSTDERR)
+	redirect_stdout(originalSTDOUT)
+
+	errstr = join(map(x->Char(x), errors))
+	outstr = join(map(x->Char(x), output))
```
Comment (omus) [inline test/interactiveutil.jl]: Can be simplified to:
```julia
errstr = UTF8String(errors)
outstr = UTF8String(output)
```

File: base/interactiveutil.jl [hunk 15]
```diff
@@ -466,9 +469,25 @@ end
...
+Note that this function relies on the availability of external tools such as `curl`, `wget`
+or `fetch` to download the file and is provided for convenience. For production use or
+situations in which more options are needed, please use a package that provides the desired
+functionality instead. 
+
+# Platform-Specific details
+
+On windows, a full URL must be provided, including the protocol prefix (e.g "http://" for
+HTTP)
+"""
```
Comment (omus) [inline base/interactiveutil.jl]: Documentation should be tied to the `download` function and not just this specific method.

File: base/interactiveutil.jl [hunk 16]
```diff
@@ -466,9 +469,25 @@ end
...
+function download(url::AbstractString; silent::Bool=false)
```
Comment (omus) [inline base/interactiveutil.jl]: Documentation should be tied to the `download` function and not just this specific method.

File: test/interactiveutil.jl (hunk 10 above)
Comment (tkelman) [inline test/interactiveutil.jl]: which does not exist on windows unless someone happens to have cygwin or msys on their path (so appveyor may not exhibit a problem)

Comment (omus) [inline test/interactiveutil.jl]: @tkelman that points out another issue. On Windows `download` uses a `ccall` which means we shouldn't be performing this check on Windows at all. Prepending `@unix_only` to this for loop should solve the issue.

File: test/interactiveutil.jl [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+end
+
+# Test silent download, no specified file
+outstr, errstr, result = check_output(download, "localhost:$port"; silent=true)
+downloaded_html = readstring(result)
+@test downloaded_html == html
+if downloader in (:curl, :wget)
+    @test length(errstr) == 0
+    @test length(outstr) == 0
+end
```
Comment (omus) [inline test/interactiveutil.jl]: Probably best if we just test that their is no output for all downloader since Windows and `fetch` are both supposed to be silent.

File: test/interactiveutil.jl [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+
+# Test verbose download, no specified file
+outstr, errstr, result = check_output(download, "localhost:$port")
+downloaded_html = readstring(result)
+@test downloaded_html == html
+if downloader in (:curl, :wget)
+    # Progress appears on stderr for curl and wget
+    @test length(errstr) > 50
+    @test length(outstr) == 0
+end
```
Comment (omus) [inline test/interactiveutil.jl]: Write an `else` test that expects there to be no output. This would cover both Windows and `fetch`

File: test/interactiveutil.jl [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+        close(errWrite)
+        close(outWrite)
+    end
+end
+
+port = 2000 #Changes to next unused port after this
+html = "<html>success</html>"
+
+function start_server()
+    global server_running, port, html
```
Comment (omus) [inline test/interactiveutil.jl]: `html` doesn't need to be global. Maybe just pass it in as an argument to `start_server`?

File: test/interactiveutil.jl [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+        errstr = UTF8String(readavailable(errRead))
+        outstr = UTF8String(readavailable(outRead))
+        # Remove the manually added \n from stdout and stderr
+        return (chomp(outstr), chomp(errstr), result)
+    catch exception
+        # Immediately restore output so errors are printed.
+        redirect_stderr(originalSTDERR)
+        redirect_stdout(originalSTDOUT)
+        println("Error checking output of $func: args=$args, kwargs=$kwargs - $exception")
```
Comment (omus) [inline test/interactiveutil.jl]: Probably should still throw an exception. I would just do `rethrow()`

**Comment by akrasian** [comment]
Fixed up the conflicts with master. Running `make julia-genstdlib` created some changes in docs/stdlib/parallel.rst and docs/stdlib/base.rst which were not on the master, so apparently a commit made it in without the .rst files being regenerated. I just reset those files to the master version, ignoring the newly generated changes, since they had nothing to do with what I was working on in order to avoid any conflict.

File: test/interactiveutil.jl [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        downloader = checkcmd
+        break
+    end
+end
+
+html = "<html>success</html>"
+start_server(html)
+
+# Test verbose download, no specified file
+outstr, errstr, result = check_output(download, "localhost:$port")
```
Comment (omus) [inline test/interactiveutil.jl]: Windows tests are probably failing since this is missing `http://` prefix

Comment (akrasian) [inline test/interactiveutil.jl]: That, and @async doesn't appear to work in windows. The code in the async block never runs, even if it's just a print statement. I'll find some other way to test in windows / disable the test in windows soon.

**Comment by omus** [comment]
@akrasian be careful to use backticks around macro names in git commit messages as it causes mentions to happen on GitHub.
